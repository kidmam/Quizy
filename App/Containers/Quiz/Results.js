// @flow

import React from 'react'
import { View, Text, StyleSheet, FlatList } from 'react-native'
import { connect } from 'react-redux'
import { NavigationActions } from 'react-navigation'
import entities from 'entities'
import { fetchQuestions, clearQuestions } from './actions'
import Button from '../../Components/Button'
import { Colors, Fonts } from '../../Themes'
import type { QuizState } from './models'

type ResultsComponent = {
  clearQuestions: Function,
  fetchQuestions: Function,
  quiz: QuizState,
  navigation: Object
}

class Results extends React.PureComponent<ResultsComponent> {
  reset = () => {
    this.props.clearQuestions()
    this.props.fetchQuestions()
    const resetAction = NavigationActions.reset({
      index: 0,
      actions: [NavigationActions.navigate({ routeName: 'Intro' })]
    })
    this.props.navigation.dispatch(resetAction)
  }

  getScore = () => this.props.quiz.results.filter((result) => result.isCorrect)

  renderItem ({ item }) {
    return (
      <View style={[styles.questionContainer, item.isCorrect ? styles.questionContainerCorrect : null]}>
        <Text style={styles.question}>{entities.decodeHTML(item.question)}</Text>
        <Text style={styles.answer}>{item.correctAnswer ? 'TRUE' : 'FALSE'}</Text>
      </View>
    )
  }

  render () {
    const score = this.getScore()
    return (
      <View style={styles.container}>
        <View style={styles.welcomeContainer}>
          <Text style={styles.welcome}>Your scored</Text>
          <Text style={styles.welcome}>
            {score.length} / {this.props.quiz.results.length}
          </Text>
          <FlatList
            keyExtractor={(_, index) => `result__${index}`}
            data={this.props.quiz.results}
            renderItem={this.renderItem}
          />
        </View>
        <Button text='PLAY AGAIN' onPress={this.reset} />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.background
  },
  welcomeContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    paddingTop: 20,
    paddingBottom: 10
  },
  welcome: {
    ...Fonts.style.h1,
    color: Colors.textAccent,
    textAlign: 'center'
  },
  questionContainer: { backgroundColor: Colors.red, paddingVertical: 10, marginTop: 5, paddingHorizontal: 30 },
  question: {
    ...Fonts.style.normal,
    color: Colors.textAccent
  },
  answer: {
    ...Fonts.style.normal,
    marginTop: 10,
    fontWeight: 'bold',
    color: Colors.textAccent,
    textAlign: 'center'
  },
  questionContainerCorrect: {
    backgroundColor: Colors.green
  }
})

const mapStateToProps = (state) => ({ quiz: state.quiz })

const mapDispatchToProps = (dispatch) => ({
  fetchQuestions: () => dispatch(fetchQuestions()),
  clearQuestions: () => dispatch(clearQuestions())
})

export default connect(mapStateToProps, mapDispatchToProps)(Results)
