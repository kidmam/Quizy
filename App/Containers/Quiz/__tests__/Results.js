import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import Results from '../Results'
import createStore from '../../../Store/Redux'

describe('Quiz Container', () => {
  let store

  beforeEach(() => {
    store = createStore(false)
  })

  it('Renders Results', () => {
    const ResultsComponent = renderer.create(<Results store={store} />)
    expect(ResultsComponent).toMatchSnapshot()
  })
})
