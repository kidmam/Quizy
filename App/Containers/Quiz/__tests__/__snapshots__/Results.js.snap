// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz Container Renders Results 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#7D59C8",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingTop": 20,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 38,
          "textAlign": "center",
        }
      }
    >
      Your scored
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 38,
          "textAlign": "center",
        }
      }
    >
      0
       / 
      0
    </Text>
    <RCTScrollView
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      invertStickyHeaders={undefined}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#4B8EFF",
        "opacity": 1,
        "paddingVertical": 20,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 30,
          "textAlign": "center",
        }
      }
    >
      PLAY AGAIN
    </Text>
  </View>
</View>
`;
